internal/controller/factory/statefulset.go

```diff
@@ -482,3 +374,269 @@ func mergeWithDefaultProbe(probe *corev1.Probe, defaultProbe corev1.Probe) *core
+		}
+		if cluster.Spec.Security.TLS.ServerSecret != "" {
+			addSecretVolume(serverCertificate.Name(), cluster.Spec.Security.TLS.ServerSecret)
+		}
+		if cluster.Spec.Security.TLS.ClientSecret != "" {
+			addSecretVolume(clientTrustedCACertificate.Name(), cluster.Spec.Security.TLS.ClientTrustedCASecret)
+		}
+	}
+
+	return volumesMap
+}
+
+func createSecretVolume(name, secretName string) corev1.Volume {
+	return corev1.Volume{
+		Name: name,
+		VolumeSource: corev1.VolumeSource{
+			Secret: &corev1.SecretVolumeSource{
+				SecretName: secretName,
+			},
+		},
+	}
+}
+
+func generateDataVolume(cluster *etcdaenixiov1alpha1.EtcdCluster) corev1.Volume {
+	var dataVolumeSource corev1.VolumeSource
+
+	if cluster.Spec.Storage.EmptyDir != nil {
+		dataVolumeSource = corev1.VolumeSource{EmptyDir: cluster.Spec.Storage.EmptyDir}
+	} else {
+		dataVolumeSource = corev1.VolumeSource{
+			PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
+				ClaimName: GetPVCName(cluster),
+			},
+		}
+	}
+
+	return corev1.Volume{
+		Name:         etcdConfig.VolumeName(),
+		VolumeSource: dataVolumeSource,
+	}
+}
+
+func updateOrAddDataVolumeMount(volumeMounts []corev1.VolumeMount,
+	volumeName, mountPath string, readOnly bool) []corev1.VolumeMount {
+	found := false
+
+	for i := range volumeMounts {
+		if volumeMounts[i].Name == volumeName {
+			volumeMounts[i].ReadOnly = readOnly
+			volumeMounts[i].MountPath = mountPath
+			found = true
+			break
+		}
+	}
+
+	if !found {
+		volumeMounts = append(volumeMounts, corev1.VolumeMount{
+			Name:      volumeName,
+			ReadOnly:  readOnly,
+			MountPath: mountPath,
+		})
+	}
+
+	return volumeMounts
+}
+
+func generateTLSSecretVolumeMounts(cluster *etcdaenixiov1alpha1.EtcdCluster) []corev1.VolumeMount {
+	volumeMounts := []corev1.VolumeMount{}
+
+	if hasTLSConfigInSpec(cluster, "") {
+		return volumeMounts
+	}
+
+	if hasTLSConfigInSpec(cluster, "PeerSecret") {
+		volumeMounts = append(volumeMounts, []corev1.VolumeMount{
+			{
+				Name:      peerTrustedCACertificate.Name(),
+				ReadOnly:  true,
+				MountPath: peerTrustedCACertificate.MountPath(),
+			},
+			{
+				Name:      peerCertificate.Name(),
+				ReadOnly:  true,
+				MountPath: peerCertificate.MountPath(),
+			},
+		}...)
+	}
+
+	if hasTLSConfigInSpec(cluster, "ServerSecret") {
+		volumeMounts = append(volumeMounts, corev1.VolumeMount{
+			Name:      serverCertificate.Name(),
+			ReadOnly:  true,
+			MountPath: serverCertificate.MountPath(),
+		})
+	}
+
+	if hasTLSConfigInSpec(cluster, "ClientSecret") {
+		volumeMounts = append(volumeMounts, corev1.VolumeMount{
+			Name:      clientTrustedCACertificate.Name(),
+			ReadOnly:  true,
+			MountPath: clientTrustedCACertificate.MountPath(),
+		})
+	}
+
+	return volumeMounts
+}
+
+func mergeVolumeMounts(lists ...[]corev1.VolumeMount) []corev1.VolumeMount {
+	var volumeMounts []corev1.VolumeMount
+
+	for _, list := range lists {
+		volumeMounts = append(volumeMounts, list...)
+	}
```

Comment:
I'd suggest to use `copy`:

```go
volumeMounts := make([]corev1.VolumeMount, len(list))
copy(volumeMounts, list)
```

Reply:
Or simply back on line 196
```go
volumeMounts := append(tlsVolumeMounts, dataVolumeMount...)
```

Reply:
This function is used in the only place, function which accepts slice as an argument and merges this slice with computed slice in its body.
IMO, it makes no sense to write dedicated function for merging slices of concrete type which accepts variadic number of arguments if it is used in one place where the number of arguments which will be passed to that function is fixed.

Reply:
> It won't work that way because the function mergeVolumeMounts accepts `...[]corev1.VolumeMount` as an argument, which allows passing N slices. While it's possible to calculate the total length of the slices, that would require a loop. In this case, using copy versus append would make no difference.

Yes, my bad. Missed that the argument is not a slice, but the slice of slices

Reply:
> Initially, there were no certificates, but then they appeared. Now, let's say some config maps and who knows what else will appear later. How would the code look then? Right now, it's very simple: we just add slices to `volumeMounts := mergeVolumeMounts(tlsVolumeMounts, dataVolumeMount, configMapVolumeMount, etcVolumeMount)`.

In case of any volume mounts defined by the user, I would expect is that function `generateVolumeMounts` will accept as arguments:
1. predefined mounts - data, tls certs, whatnot else
1. container volumeMounts defined in cluster spec by the user

and merge these two lists. Computation of predefined mounts is the separate step where we need to take into account provided cluster configuration, like security config.
Bam, profit.